### PR TITLE
Bug 1539232 - Switch Perfherder to react-router

### PR DESCRIPTION
**Pull request by sarah-clements**
This is the final react conversion pr - **woohoo!**

In addition to setting up the routes, I've made these other changes:
- Using top-level of app as a cache for projects, frameworks, alerts data and compare data
- Cleanup files and move constants to dedicated perfherder file
- Removed angular-related libraries and bump down the neutrino entry and asset limits

@ionutgoldan it'd be worth you giving all the views a quick test. I unfortunately had a lot of squashing of commits to do during the rebase and I want to make sure the new retrigger action is working as expected. I'd also like to ensure that the fetched data that's stored from the alerts view (when navigating back from an individual alert), and the fetched data that's stored from the compare view (when navigating back from the comparesubtest view) works as expected and doesn't cause any issues. I've launched it to prototype.

FYI @karlht and @davehunt

### CODE REVIEW

**Comment by sarah-clements** [comment]
@camd There's a selenium test that's failing - the [test](https://github.com/mozilla/treeherder/blob/master/tests/selenium/test_switch_app.py) for switching between Treeherder and Perfherder in the dropdown menu. Do you think this is worth trying to get working or were you planning to rewrite at some point? I've rewritten the other tests for Perfherder using react-testing-library so this is the only remnant of Perfherder selenium tests.

**Comment by sarah-clements** [comment]
> @camd There's a selenium test that's failing - [this test](https://github.com/mozilla/treeherder/blob/master/tests/selenium/test_switch_app.py) for switching between Treeherder and Perfherder in the dropdown menu. Do you think this is worth trying to get working or were you planning to rewrite at some point? I've rewritten the other tests for Perfherder using react-testing-library so this is the only remnant of Perfherder selenium tests.

Cameron's #5383 pr fixed this problem.

**Comment by ionutgoldan** [comment]
I only managed to see that the permalinks from the Compare view no longer work as expected.
Prior to this PR, the clicking on a permalink would put the item at the top of the page. Now, that item is no longer visible.

**Comment by ionutgoldan** [comment]
@alexandru-io @bebef1987 could you help me out on testing this PR next Monday?
If we have more eyes on this, we can spot more bugs.

**Comment by sarah-clements** [comment]
> I only managed to see that the permalinks from the Compare view no longer work as expected.
> Prior to this PR, clicking on a permalink would put the item at the top of the page. Now, that item is no longer visible.
> 
> **STR**
> 
>     1. Go to a Compare view with some data ([e.g.](https://treeherder-prototype.herokuapp.com/perf.html#/compare?originalProject=mozilla-central&originalRevision=a3a081ae714f1123bdc23c9d9ef53dfaa783a8de&newProject=mozilla-central&newRevision=6b93a83735ed3ab3b57b46c1b768814b1a1af5d6&framework=10))
> 
>     2. Click on the [# permalink](https://treeherder-prototype.herokuapp.com/perf.html#/compare?originalProject=mozilla-central&originalRevision=a3a081ae714f1123bdc23c9d9ef53dfaa783a8de&newProject=mozilla-central&newRevision=6b93a83735ed3ab3b57b46c1b768814b1a1af5d6&framework=10#table-header-566498077) next to `raptor-ares6-chromium opt`.
> 
> 
> Expected:
> The `raptor-ares6-chromium opt` header should show up at the top of the page.
> 
> What we actually get:
> `raptor-ares6-chromium opt` header isn't visible anymore. At least, not entirely.

Ah, I know why it's doing that. It's because of how the top menu navigation is overlapping the view. So the permalink is still working correctly, I'll just need to fix how the menu is displayed.

File: ui/perfherder/compare/CompareTable.jsx [hunk 1]
```diff
@@ -33,7 +32,7 @@ export default class CompareTable extends React.PureComponent {
       displayNumber(percentage),
     )}% ${improvement ? 'better' : 'worse'})`;
 
-  // humane readable signature name
```
Comment (camd) [inline ui/perfherder/compare/CompareTable.jsx]: LOL...

**Comment by ionutgoldan** [comment]
I am now resuming the testing on this PR.

**Comment by ionutgoldan** [comment]
`Alerts` navbar item has weird behavior.

**STR**
1. Go to default landing view of Perfherder. *(which is Alerts view)*
2. Click on `Compare` navbar item.
3. After it loads, click on `Alerts` navbar item & wait until it fully loads.
4. Click on `Alerts` navbar item one more time.

Expected results:
A page similar to that from step #3. That is a page with alerts summaries

What we actually get:
A page with no alert summaries, which states "No alerts to show".

This bug is specific to this PR.

**Comment by ionutgoldan** [comment]
**Bug #3**
Permalinks no longer visible in Compare subtest view

**STR**
1. Go to a `Compare` view with data. ([e.g.](https://treeherder-prototype.herokuapp.com/perf.html#/compare?originalProject=mozilla-central&originalRevision=67ba5406ce3954311b41ea6bb5564118a5699244&newProject=mozilla-inbound&newRevision=19c33f741d779768f53eed13e5c0f45e0363f537&framework=12))
2. Follow one of the available `subtests` hyperlinks. ([e.g.](https://treeherder-prototype.herokuapp.com/perf.html#/comparesubtest?originalProject=mozilla-central&newProject=mozilla-inbound&newRevision=19c33f741d779768f53eed13e5c0f45e0363f537&originalSignature=1943598&newSignature=1949646&framework=12&originalRevision=67ba5406ce3954311b41ea6bb5564118a5699244))

Expected results:
Table header & all subtests display a `#` permalinks.

What we actually get:
No `#` permalinks are displayed.

**Comment by ionutgoldan** [comment]
**Bug #4**
Retrigger buttons from `Compare` view & `Compare` subtest view are broken.

**STR**
1. Go to `Compare` view with data. ([e.g.](https://treeherder-prototype.herokuapp.com/perf.html#/compare?originalProject=mozilla-central&originalRevision=67ba5406ce3954311b41ea6bb5564118a5699244&newProject=mozilla-central&newRevision=2b8364cfdb04bfdfde7c01589e48e24973c1ce9c&framework=10))
2. Click on the retrigger button under the `# Runs` column from right side, from the `raptor-ares6-chromium opt` table, the `linux64-shippable` row.

Expected results:
No console errors & new AC(rt) jobs are pending for [base](https://treeherder-prototype.herokuapp.com/#/jobs?repo=mozilla-central&selectedJob=267108483&group_state=expanded&revision=67ba5406ce3954311b41ea6bb5564118a5699244&searchStr=ac%28rt%29) & [new](https://treeherder-prototype.herokuapp.com/#/jobs?repo=mozilla-central&group_state=expanded&revision=2b8364cfdb04bfdfde7c01589e48e24973c1ce9c&searchStr=action) on Treeherder.

What we actually get:
`TypeError: s is not a function` console error & no retriggers are done Treeherder.

**Comment by ionutgoldan** [comment]
**Bug #5**
Pagination from bottom of page still visible when checking out a single summary.

**STR**
1. Go to random `Alerts` view. ([e.g.](https://treeherder-prototype.herokuapp.com/perf.html#/alerts?hideDwnToInv=1&framework=6))
2. Checkout a random summary from the available ones there. ([e.g.](https://treeherder-prototype.herokuapp.com/perf.html#/alerts?id=22078))

Expected results:
No pagination at the bottom of page.

What we actually get:
A gray & squared `1` page at the bottom of page.

**Comment by ionutgoldan** [comment]
**Bug #6**
`hideDwnToInv=1` now shows up in the URL of checkout out summaries.

**STR**
1. Go to random `Alerts` view. ([e.g.](https://treeherder-prototype.herokuapp.com/perf.html#/alerts?hideDwnToInv=1&framework=6))
2. Randomly look at 2 alert summaries, from the same repository. (e.g. [1st summary](https://treeherder-prototype.herokuapp.com/perf.html#/alerts?id=22109&hideDwnToInv=1), [2nd summary](https://treeherder-prototype.herokuapp.com/perf.html#/alerts?id=22106&hideDwnToInv=1))
3. Reassigned all alerts from the 1st summary to the 2nd summary.
4. Checkout to the 1st alert summary page. ([e.g.](https://treeherder-prototype.herokuapp.com/perf.html#/alerts?id=22109&hideDwnToInv=1))

Expected results:
The `Hide downstream / reassigned to / invalid` checkbox isn't enabled, so that user sees all alerts stroke through.

What we actually get:
The `Hide downstream / reassigned to / invalid` is enabled.

**Comment by marianrai** [comment]
Login issue on Firefox and Chrome;

STR
1. Go to https://treeherder-prototype.herokuapp.com/perf.html
2. Logged in successfully
3. After step2 i see a blue alert info on the top : "You must be logged into perfherder/treeherder and be a sheriff to make changes". On top right i can see my name, but i'm not really logged in.
4. I can't use the checkboxes on alert summaries nor open new alerts from graphs.
5. Cleared cache, cookies, restarted browsers - issue persists.

Expected results:
After the login i should have full functionality on perfherder.

What we get instead:
Perfherder acts as i am not logged in.
In Chrome i can see this error in console:
"Login issuie in heroku perfherder
AuthService.js:66 Could not renew login: Object
_renewAuth @ AuthService.js:66"

**Comment by ionutgoldan** [comment]
> 
> 
> Login issue on Firefox and Chrome;
> 
> STR
> 
>     1. Go to https://treeherder-prototype.herokuapp.com/perf.html
> 
>     2. Logged in successfully
> 
>     3. After step2 i see a blue alert info on the top : "You must be logged into perfherder/treeherder and be a sheriff to make changes". On top right i can see my name, but i'm not really logged in.
> 
>     4. I can't use the checkboxes on alert summaries nor open new alerts from graphs.
> 
>     5. Cleared cache, cookies, restarted browsers - issue persists.
> 
> 
> Expected results:
> After the login i should have full functionality on perfherder.
> 
> What we get instead:
> Perfherder acts as i am not logged in.
> In Chrome i can see this error in console:
> "Login issuie in heroku perfherder
> AuthService.js:66 Could not renew login: Object
> _renewAuth @ AuthService.js:66"
> 
> Video with the issue from Alex: https://mozilla.zoom.us/recording/play/g5UjN_GDTxhejx0bepu3XyRSFa94S4_-wqFQyY652kSrTXPwMeqzHTgwp0ke4c7T?continueMode=true

This is a false alarm. You were missing the `is_staff` flag set on the prototype database. I've addressed that and you should no longer experience this issue.

**Comment by sarah-clements** [comment]
> **Bug #2**
> `Alerts` navbar item has weird behavior, on the 2nd click.
> 
> **STR**
> 
>     1. Go to default landing view of Perfherder. _(which is Alerts view)_
> 
>     2. Click on `Compare` navbar item, from top-left.
> 
>     3. After it loads, click on `Alerts` navbar item & wait until it fully loads.
> 
>     4. Click on `Alerts` navbar item one more time.
> 
> 
> Expected results:
> A page similar to that from step #3. That is a page with alerts summaries
> 
> What we actually get:
> A page with no alert summaries, which states "No alerts to show".
> 
> This bug is specific to this PR.

Ok, interesting. Why would you click on alerts twice though? Is that a typical work flow?

**Comment by sarah-clements** [comment]
@ionutgoldan Btw, when you write `#<some number>` it links to old Treeherder pull requests and shows up as referencing that bug (click on your `Bug #2` for example so it'd be best to use different notation :)

**Comment by sarah-clements** [comment]
Thanks for the feedback @ionutgoldan - I'll continue working on these fixes tomorrow and give it a good last test before merging.

**Comment by ionutgoldan** [comment]
> 
> 
> @ionutgoldan Btw, when you write `#<some number>` it links to old Treeherder pull requests and shows up as referencing that bug (click on your `Bug #2` for example so it'd be best to use different notation :)

Ok. Will use a different notation from now on. Thanks!

**Comment by alexandru-io** [comment]
Bug # 8
Data points for new test added in graph with button **Add test data** drawn with white.

STR

1. Click the graph button of an alert
2. Click **Add test data** and select the same test but on another platform (linux-64-shippable in this case)
3. See that the datapoints are drawn white. ([e.g.](https://treeherder.mozilla.org/perf.html#/graphs?timerange=1209600&series=autoland,1922714,1,1&series=autoland,1927614,1,1&highlightAlerts=1))

Expected results:
The data points of the added hould be drawn in a different color than the background of the graph.

What we actually get:
The data points of the added test are drawn with the color of graph background.

**Comment by sarah-clements** [comment]
> **Bug #5**
> Pagination from bottom of page still visible when checking out a single summary.
> 
> **STR**
> 
>     1. Go to random `Alerts` view. ([e.g.](https://treeherder-prototype.herokuapp.com/perf.html#/alerts?hideDwnToInv=1&framework=6))
> 
>     2. Checkout a random summary from the available ones there. ([e.g.](https://treeherder-prototype.herokuapp.com/perf.html#/alerts?id=22078))
> 
> 
> Expected results:
> No pagination at the bottom of page.
> 
> What we actually get:
> A gray & squared `1` page at the bottom of page.
> 
> _Note:_ this is a preexisting bug, which I've found just now.

I don't see this as a bug per say - technically it is the the one and only page. However, if you would like to improve upon the pagination please feel free.

**Comment by davehunt** [comment]
> Note: this is a preexisting bug, which I've found just now.

Let's file any pre-existing bugs in bugzilla and not block this patch on them.

**Comment by sarah-clements** [comment]
> Bug # 8
> Data points for new test added in graph with button **Add test data** drawn with white.
> 
> STR
> 
>     1. Click the graph button of an alert
> 
>     2. Click **Add test data** and select the same test but on another platform (linux-64-shippable in this case)
> 
>     3. See that the datapoints are drawn white. ([e.g.](https://treeherder.mozilla.org/perf.html#/graphs?timerange=1209600&series=autoland,1922714,1,1&series=autoland,1927614,1,1&highlightAlerts=1)) [Recording here](https://mozilla.zoom.us/recording/play/fTDrKg1yPwMyfqyz0Z6gAgI1kw1xq_upbPss9CrUfQ6VKJef6gqpOm3Niy5S3iPF?continueMode=true)
> 
> 
> Expected results:
> The data points of the added should be drawn in a different color than the background of the graph.
> 
> What we actually get:
> The data points of the added test are drawn with the color of graph background.
> Note that after closing the tab and opening another one, the added test is visible. This means that once in a while the color picked up by the graph to plot is white.

Thanks for reporting this but it doesn't have anything to do with this pr, which is for testing react router changes on [prototype](https://treeherder-prototype.herokuapp.com/). Please file a bug with the video. I was able to reproduce this once - following the sequence from your video - but then wasn't able to reproduce it again. This might be an edge case. I'd be curious if it happened after the tab was left open for a long time or after frequent use of the app (maybe that makes a difference, maybe not).

**Comment by ionutgoldan** [comment]
> 
> 
> > Note: this is a preexisting bug, which I've found just now.
> 
> Let's file any pre-existing bugs in bugzilla and not block this patch on them.

Will do.

**Comment by edmorley** [comment]
@sarah-clements @camd Congrats on finishing the AngularJS migration! :-)

**Comment by camd** [comment]
Thanks, @edmorley !  Great to hear from you!  I hope you're well, old friend.  :)

**Comment by sarah-clements** [comment]
Yes, @edmorley, it was a lot of work converting Perfherder and I'm glad it's finished! You should pop in the irc channel some time before it goes away... :)

**Pull request merged**